### ISSUE TITLE
Tuple size is lost while it can be deduced

### ISSUE THREAD

**Issue by iyanging** [microsoft/pyright#6958]
**Describe the bug**
Tuple size is lost while it can be deduced.

**Code or Screenshots**

```python
def hello(a: int) -> tuple[int, int, int]:
    r = (1, 2, 3, 4)
    if a > 0:
        return r[0:3]
    else:
        return r[1:4]

```

```Shell
❯ pyright a.py
/Users/iyang/a.py
  /Users/iyang/a.py:6:16 - error: Expression of type "tuple[Literal[1, 2, 3, 4], ...]" cannot be assigned to return type "tuple[int, int, int]"
    "tuple[Literal[1, 2, 3, 4], ...]" is incompatible with "tuple[int, int, int]"
      Tuple size mismatch; expected 3 but received indeterminate (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

**VS Code extension or command-line**
```Shell
> pyright --version
pyright 1.1.345
```

**Comment by tsibley** [comment]
I've also run into this where a tuple typed `Tuple[str, int] | Tuple[str, int, int, int]` sliced with `[0:2]` results in the revealed type `tuple[str | int, ...]` instead of `tuple[str, int]`.

**Comment by erictraut** [comment]
Thanks for the bug report. This was an off-by-one error in my logic — and a hole in (what was otherwise an extensive) test suite. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.346, which I just published. It will also be included in a future release of pylance.
